// file-graph-shared/src/file.js
const $ = require('jquery');
const _ = require('lodash');

class File {
    constructor(path, name, tags = []) {
        this.path = path;
        this.name = name;
        this.tags = tags;
        this.listeners = [];
    }

    getId() {
        return this.path + '/' + this.name;
    }

    save() {
        return new Promise((res, rej) => {
            $.ajax({
                type: 'POST',
                url: `/file?file=${JSON.stringify(this.toJSON())}`,
                dataType: 'json',
                success: res,
                contentType: 'application/json'
            });
        })
            .then((data) => {
                this.listeners.forEach(l => l.onFileUpdate(data));
            })
            .then(() => this);
    }

    addListener(listener) {
        this.listeners.push(listener);
    }

    static fromJSON(json) {
        return new File(json.path, json.name, json.tags);
    }

    toJSON() {
        return _.pick(this, ['path','name','tags'])
    }
}

module.exports = File;
