// file-graph-shared/src/model/edge.js
const ListenableObject = require('../listenable-object');

class Edge extends ListenableObject {
    constructor(leftNode, rightNode, { tags = [], constantAttributes = {} } = {}) {
        super();
        this.id = Edge.generateId(leftNode, rightNode);
        this.leftNode = leftNode;
        this.rightNode = rightNode;
        this.tags = tags;
        this.constantAttributes = constantAttributes;
        this.set('highlighted', false);
        this.set('selected', false);

        this.state = new ListenableObject({
            hasHighlightedNode: false,
            hasSelectedNode: false
        });

        this.listenTo(this.leftNode, 'change:selected', this.onNodeSelectionChanged);
        this.listenTo(this.rightNode, 'change:selected', this.onNodeSelectionChanged);
        this.listenTo(this.leftNode, 'change:highlighted', this.onNodeSelectionChanged);
        this.listenTo(this.rightNode, 'change:highlighted', this.onNodeSelectionChanged);
    }

    onNodeSelectionChanged() {
        const hasHighlightedNode =
            this.leftNode.get('highlighted') || this.rightNode.get('highlighted');
        const hasSelectedNode =
            this.leftNode.get('selected') || this.rightNode.get('selected');
        this.set('hasHighlightedNode', hasHighlightedNode);
        this.set('hasSelectedNode', hasSelectedNode);
    }

    addTag(tag) {
        this.tags.push(tag);
    }

    addNode(node) {
        this.nodes[node.id] = node;
    }

    removeNode(node) {
        delete this.nodes[node.id];
    }

    get selected() {
        return this.get('selected');
    }

    set selected(isSelected) {
        this.set('selected', isSelected);
    }

    get highlighted() {
        return this.get('highlighted');
    }

    set highlighted(isHighlighted) {
        this.set('highlighted', isHighlighted);
    }

    removeListeners() {
        this.stopListening(this.leftNode, 'change:selected');
        this.stopListening(this.rightNode, 'change:selected');
        this.stopListening(this.leftNode, 'change:highlighted');
        this.stopListening(this.rightNode, 'change:highlighted');
    }

    static generateId(node1, node2) {
        const nodeIds = [node1.id, node2.id];
        return nodeIds.sort().join('-');
    }
}

module.exports = Edge;
