// file-graph-shared/src/model/staticAttribute.js
const _ = require('lodash');

const ListenableObject = require('../listenable-object');


class StaticAttribute extends ListenableObject {
    constructor(group, value, force) {
        super();
        this.id = group + value;
        this.group = group;
        this.value = value;
        this._nodes = {};
        this.set('force', force);
        this.set('highlighted', false);
    }

    get nodes() {
        return _.values(this._nodes);
    }

    get highlighted() {
        return this.get('highlighted');
    }

    set highlighted(isHighlighted) {
        this.set('highlighted', isHighlighted);
    }

    addNode(node) {
        this._nodes[node.id] = node;
    }

    removeNode(node) {
        delete this._nodes[node.id];
        this.emitEvent('removed:node', node);
    }

    toJSON() {
        return {
            group: this.group,
            value: this.value,
            force: this.get('force'),
        };
    }

    static fromJSON({ group, value }) {
        return new StaticAttribute(group, value);
    }
}

module.exports = StaticAttribute;
