// file-graph-shared/src/model/tag.js
const _ = require('lodash');

const ListenableObject = require('../listenable-object');


class Tag extends ListenableObject {
    constructor(name) {
        super();
        this.id = name;
        this._nodes = {};
        this.set('name', name);
        this.set('highlighted', false);
        this.set('selected', false);
    }

    get nodes() {
        return _.values(this._nodes);
    }

    get name() {
        return this.get('name');
    }

    get selected() {
        return this.get('selected');
    }

    set selected(isSelected) {
        this.set('selected', isSelected);
    }

    get highlighted() {
        return this.get('highlighted');
    }

    set highlighted(isHighlighted) {
        this.set('highlighted', isHighlighted);
    }


    addNode(node) {
        this._nodes[node.id] = node;
    }

    removeNode(node) {
        delete this._nodes[node.id];
        this.emitEvent('removed:node', node);
    }

    toJSON() {
        return {
            name: this.name
        }
    }

    static fromJSON({ name }) {
        return new Tag(name);
    }
}

module.exports = Tag;
